File: trapping-rain-water.java
// https://leetcode.com/problems/trapping-rain-water/description/
class Solution {
    // Dynamic Programming:
    // 1. Find the water horizon: save the max numbers right->left, left->right, and the min numbers among them
    // 2. Subtract height - water horizon
    // Time: O(n), Space: O(n)
    public int trap(int[] height) {
        int l = height.length; if(l==0) return 0;
        int[] max = new int[l];
        int t = 0;
        for(int i=(l-1); i>=0; i--){
            t = Math.max(t, height[i]);
            max[i] = t;
        }
        t = 0;
        int ans = 0;
        for(int i=0; i<l; i++){
            t = Math.max(t, height[i]);
            max[i] = Math.min(max[i], t);
            ans += max[i] - height[i];
        }
        return ans;
    }
    // Two pointers (left, right)
    // left, leftMax <------------> right, rightMax
    // increase by height & get water horizon & subtract height - water horizon
    // Time: O(n), Space: O(1)
    public int trap2(int[] height) {
        int lMax = 0, rMax = 0, l = 0, r = height.length-1, ans = 0;
        while(l<r){
            //System.out.println(l+" "+r+" "+lMax+" "+rMax+" "+ans);
            if(height[l]>height[r]){
                rMax = Math.max(rMax, height[r]);
                ans += rMax - height[r];
                r--;
            }else{
                lMax = Math.max(lMax, height[l]);
                ans += lMax - height[l];
                l++;
            }
            //System.out.println(" "+l+" "+r+" "+lMax+" "+rMax+" "+ans);
        }
        return ans;
    }
}

File: trapping-rain-water.py
# https://leetcode.com/problems/trapping-rain-water/description/
class Solution:
    # two pointers (left, right)
    # Time: O(n), Space: O(1)
    def trap(self, height):
        ans = 0
        l,r = 0,len(height)-1
        maxL,maxR = 0,0
        while l<r:
            if height[l] < height[r]:
                if height[l] >= maxL: maxL = height[l]
                else: ans += maxL - height[l]
                l += 1
            else:
                if height[r] >= maxR: maxR = height[r]
                else: ans += maxR - height[r]
                r -= 1
        return ans
    
    # stack (save decreasing or the same, and pop when increasing)
    # Time: O(n), Space: O(n)
    def trap3(self, height):
        ans = 0
        stk = []
        for i in range(len(height)):
            while stk and height[i]>height[stk[-1]]:
                j = stk.pop()
                if not stk: break
                else: ans += (i-stk[-1]-1)*(min(height[i],height[stk[-1]])-height[j])
            stk.append(i)
        return ans            
    
    # 2-pass: get min(Max from left, Max from right)
    # Time: O(n), Space: O(n)
    def trap2(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        l = len(height)
        waterH = [0 for i in range(l)]
        
        # left -> right
        maxL = 0
        for i in range(l):
            maxL = max(maxL, height[i])
            waterH[i] = maxL
        
        # right -> left
        # and get the ans
        maxR = 0
        ans = 0
        for i in range(l-1,-1,-1):
            maxR = max(maxR, height[i])
            ans += min(maxR, waterH[i]) - height[i]
        
        return ans
